metrics/api/src/main/java/io/helidon/metrics/api/AbstractRegistry.java

```diff
@@ -0,0 +1,808 @@
+    }
+
+    @Override
+    public ConcurrentGauge concurrentGauge(String name, Tag... tags) {
+        return getOrRegisterMetric(name, ConcurrentGauge.class, tags);
+    }
+
+    @Override
+    public ConcurrentGauge concurrentGauge(Metadata metadata, Tag... tags) {
+        return getOrRegisterMetric(metadata, ConcurrentGauge.class, tags);
+    }
+
+    @Override
+    public SimpleTimer simpleTimer(String name) {
+        return simpleTimer(name, NO_TAGS);
+    }
+
+    @Override
+    public SimpleTimer simpleTimer(Metadata metadata) {
+        return simpleTimer(metadata, NO_TAGS);
+    }
+
+    @Override
+    public SimpleTimer simpleTimer(String name, Tag... tags) {
+        return getOrRegisterMetric(name, SimpleTimer.class, tags);
+    }
+
+    @Override
+    public SimpleTimer simpleTimer(Metadata metadata, Tag... tags) {
+        return getOrRegisterMetric(metadata, SimpleTimer.class, tags);
+    }
+
+    /**
+     * Removes a metric by name. Synchronized for atomic update of more than one internal map.
+     *
+     * @param name Name of the metric.
+     * @return Outcome of removal.
+     */
+    @Override
+    public synchronized boolean remove(String name) {
+        final List<Map.Entry<MetricID, M>> doomedMetrics = getMetricsByName(name);
+        if (doomedMetrics.isEmpty()) {
+            return false;
+        }
+
+        final boolean result = doomedMetrics.stream()
+                .peek(entry -> entry.getValue().markAsDeleted())
+                .map(entry -> allMetrics.remove(entry.getKey()) != null)
+                .reduce((a, b) -> a || b)
+                .orElse(false);
+        allMetricIDsByName.remove(name);
+        allMetadata.remove(name);
+        return result;
+    }
+
+    /**
+     * Removes a metric by ID. Synchronized for atomic update of more than one internal map.
+     *
+     * @param metricID ID of metric.
+     * @return Outcome of removal.
+     */
+    @Override
+    public synchronized boolean remove(MetricID metricID) {
+        final List<MetricID> metricIDS = allMetricIDsByName.get(metricID.getName());
+        if (metricIDS == null) {
+            return false;
+        }
+        metricIDS.remove(metricID);
+        if (metricIDS.isEmpty()) {
+            allMetricIDsByName.remove(metricID.getName());
+            allMetadata.remove(metricID.getName());
+        }
+
+        M doomedMetric = allMetrics.remove(metricID);
+        if (doomedMetric != null) {
+            doomedMetric.markAsDeleted();
+        }
+        return doomedMetric != null;
+    }
+
+    @Override
+    public void removeMatching(MetricFilter filter) {
+        allMetrics.entrySet().stream()
+                .filter(entry -> filter.matches(entry.getKey(), entry.getValue()))
+                .map(entry -> remove(entry.getKey()))
+                .reduce((a, b) -> a || b);
+    }
+
+    @Override
+    public SortedSet<String> getNames() {
+        return allMetrics.keySet().stream()
```

Comment:
Would `allMetricIDsByName.keySet()` work here?

Reply:
Not quite. The keys in the `allMetrics` `Map` are not sorted as the return value must be. 

But I will change it to
```
return new TreeSet<>(allMetricsIDsByName.keySet());
```
which lets the constructor decide how it wants to build the `TreeSet` instead of our code choosing how.

Reply:
Don't forget to wrap it in `Collections.unmodifiableSortedSet()`.

Reply:
Hmm. The contract for `getNames` in the MP Metrics `MetricRegistry` interface doesn't indicate that the result will be immutable. Plus, our prior implementation returns a mutable set, so although you might have a good point returning an immutable instance would be a potentially backward-incompatible change.